### PR TITLE
[circt-bmc] Add simple initial value support to ExternalizeRegisters

### PR DESCRIPTION
**Pull request by TaoBi22**
Adds support for initial values (as long as they're defined directly by a `hw.constant` op inside a `seq.initial`) to `ExternalizeRegisters` - I'll add the support for this in circt-bmc as a follow-up to slim down the diff. Initial values are added in an ArrayAttr on the module op - if there's no initial value then it just adds a unit attr

### CODE REVIEW

File: lib/Tools/circt-bmc/ExternalizeRegisters.cpp [hunk 1]
```diff
@@ -93,9 +94,39 @@ void ExternalizeRegistersPass::runOnOperation() {
...
           }
-          if (regOp.getInitialValue()) {
-            regOp.emitError("registers with initial values not yet supported");
-            return signalPassFailure();
+          mlir::Attribute initState;
+          if (auto initVal = regOp.getInitialValue()) {
+            // Find the seq.initial op where the initial value is defined and
+            // fetch the operation inside that defines the value
+            if (isa<BlockArgument>(initVal) ||
+                !isa<seq::InitialOp>(initVal.getDefiningOp())) {
```
Comment (maerhart) [inline lib/Tools/circt-bmc/ExternalizeRegisters.cpp:102]:
```suggestion
            auto initialOp = initVal.getDefiningOp<seq::InitialOp>();
            if (!initialOp) {
```

File: lib/Tools/circt-bmc/ExternalizeRegisters.cpp [hunk 2]
```diff
@@ -93,9 +94,39 @@ void ExternalizeRegistersPass::runOnOperation() {
...
+              regOp.emitError("registers with initial values not directly "
+                              "defined by a seq.initial op not yet supported");
+              return signalPassFailure();
+            }
+            auto *initialOp = regOp.getInitialValue().getDefiningOp();
+            auto index =
+                llvm::find(initialOp->getResults(), initVal).getIndex();
```
Comment (maerhart) [inline lib/Tools/circt-bmc/ExternalizeRegisters.cpp:109]:
```suggestion
            auto index = cast<OpResult>(initVal).getResultNumber();
```

File: lib/Tools/circt-bmc/ExternalizeRegisters.cpp [hunk 3]
```diff
@@ -93,9 +94,39 @@ void ExternalizeRegistersPass::runOnOperation() {
...
+            auto initValDef =
+                initialOp->getRegion(0).front().getTerminator()->getOperand(
+                    index);
+            // If it's defined by a constant op then just fetch the constant
+            // value - otherwise unsupported
+            if (auto constantOp =
+                    dyn_cast<hw::ConstantOp>(initValDef.getDefiningOp())) {
```
Comment (maerhart) [inline lib/Tools/circt-bmc/ExternalizeRegisters.cpp:116]:
```suggestion
                    initValDef.getDefiningOp<hw::ConstantOp>()) {
```

**Review by maerhart — APPROVED** [review]
Very cool! Thanks for adding that! 🎉

File: lib/Tools/circt-bmc/ExternalizeRegisters.cpp (hunk 1 above)
Comment (TaoBi22) [reply lib/Tools/circt-bmc/ExternalizeRegisters.cpp:102]: Nice, thanks!

**Pull request merged**